### PR TITLE
transform.blur

### PR DESCRIPTION
**Pull request by yunline**
# pygame.transform.blur

*This PR is cloned from [https://github.com/pygame/pygame/pull/3698](https://github.com/pygame/pygame/pull/3698).*

**2023-02-08T02:53:34Z** [yunline](https://github.com/yunline) wrote:
> # Add `blur` to transform module.
  > ```python
  > def box_blur(
  >     surface: Surface,
  >     radius: int,
  >     repeat_edge_pixels: Optional[bool] = True,
  >     dest_surface: Optional[Surface] = None
  > ) -> Surface: ...
  > 
  > def gaussian_blur( ... ) -> Surface: ...
  > ```
  > Inspired by #3659.
  > 
  > image
  > ☑ Box Blur
  > ☑ Gaussian Blur

**2023-02-04T23:21:21Z** [novialriptide](https://github.com/novialriptide) wrote:
> Looks like you need to run `python setup.py format` to pass the linting test.  

**2023-02-06T01:41:11Z** [novialriptide](https://github.com/novialriptide) wrote:
> Blurring Better Call Saul is giving me weird results here... I'm not sure if this is normal
  > 
  > ### Result
  > image
  > 
  > ### Original Image
  > maxresdefault
  > <details>
  > <summary>3d</summary>
  > 
  > saul-goodman-better-call-saul
  > 
  > </details>  

**2023-02-06T02:10:51Z** [yunline](https://github.com/yunline) wrote:
> > Blurring Better Call Saul is giving me weird results here... I'm not sure if this is normal
  > 
  > Fixed now, thanks for your testing.
  > image
  >   

**2023-02-06T03:52:35Z** [novialriptide](https://github.com/novialriptide) wrote:
> You can never go wrong with Saul's good looks. :)  

**2023-02-06T08:16:58Z** [yunline](https://github.com/yunline) wrote:
> > It takes about 220ms to process an image using this algorithm.
  > 
  > I optimized the algorithm and now it takes 75ms to process one image.
  > 
  > Image: 2048x2048 RGBA
  > CPU: AMD Ryzen7 6800H  

**2023-02-06T10:39:46Z** [itzpr3d4t0r](https://github.com/itzpr3d4t0r) wrote:
> I think that a blur function should be named depending on the blur algorithm it uses, so gaussian, bokeh etc.  

**2023-02-06T20:18:01Z** [yunline](https://github.com/yunline) wrote:
> New parameter `repeat_edge_pixels`
  > image  

**2023-02-08T02:50:03Z** [yunline](https://github.com/yunline) wrote:
> Gaussian blur √
  > image

### CODE REVIEW

**Comment by yunline** [comment]
Relative issue #1796

**Review by Matiiss — CHANGES_REQUESTED** [review]
Looks great, some minor stuff should be fixed so the compiler doesn't complain, also docs and don't forget to run `python setup.py format` when you're done and before pushing so that the lint test can pass too.

File: buildconfig/stubs/pygame/transform.pyi [hunk 1]
```diff
@@ -51,3 +51,15 @@ def threshold(
...
+    radius: int,
+    repeat_edge_pixels: Optional[bool] = True,
+    dest_surface: Optional[Surface] = None
+) -> Surface: ...
+def gaussian_blur(
+    surface: Surface,
+    radius: int,
+    repeat_edge_pixels: Optional[bool] = True,
+    dest_surface: Optional[Surface] = None
+) -> Surface: ...
```
Comment (MyreMylar) [inline buildconfig/stubs/pygame/transform.pyi:65]: GitHub is telling me there is a missing new line here.

**Review by MyreMylar — COMMENTED** [review]
The most obvious thing missing is tests on the output of these two algorithms.

Tests would be particularly handy in this case as I expect a long term project we will embark on is to try and utilise SIMD techniques in the transform submodule in the same way we've been doing in the surface submodule. If we have tests verifying the output then a future SIMD version of these blurs would be able to test that what it is doing matches the current linear math version.

**Comment by oddbookworm** [comment]
@yunline you should run `setup.py format` to pass the lint check

File: buildconfig/stubs/pygame/transform.pyi [hunk 2]
```diff
@@ -51,3 +51,15 @@ def threshold(
     search_surf: Optional[Surface] = None,
     inverse_set: bool = False,
 ) -> int: ...
+def box_blur(
+    surface: Surface,
+    radius: int,
+    repeat_edge_pixels: Optional[bool] = True,
```
Comment (Matiiss) [inline buildconfig/stubs/pygame/transform.pyi:57]: It's just a `bool`, `Optional` means either `None` or the type given
`repeat_edge_pixels: bool = True`

File: buildconfig/stubs/pygame/transform.pyi [hunk 3]
```diff
@@ -51,3 +51,15 @@ def threshold(
...
+    dest_surface: Optional[Surface] = None
+) -> Surface: ...
+def gaussian_blur(
+    surface: Surface,
+    radius: int,
+    repeat_edge_pixels: Optional[bool] = True,
```
Comment (Matiiss) [inline buildconfig/stubs/pygame/transform.pyi:63]: Same for this with `Optional`

File: docs/reST/ref/transform.rst [hunk 4]
```diff
@@ -210,6 +210,29 @@ Instead, always begin with the original image and scale to the desired size.)
 
    .. ## pygame.transform.laplacian ##
 
+.. function:: box_blur
+
+   | :sl:`make surface blurry`
```
Comment (Matiiss) [inline docs/reST/ref/transform.rst:215]: Based on other docs it should be sth like
"blur a surface using box blur"

File: docs/reST/ref/transform.rst [hunk 5]
```diff
@@ -210,6 +210,29 @@ Instead, always begin with the original image and scale to the desired size.)
...
+
+   Returns the blured surface using box blur algorithm.
+
+   .. versionadded:: 2.1.4
+
+   .. ## pygame.transform.box_blur ##
+
+.. function:: gaussian_blur
+
+   | :sl:`make surface blurry`
```
Comment (Matiiss) [inline docs/reST/ref/transform.rst:226]: "blur a surface using gaussian blur"

**Review by Matiiss — CHANGES_REQUESTED** [review]
Couple more doc and stub changes would be great

File: buildconfig/stubs/pygame/transform.pyi (hunk 2 above)
Comment (yunline) [reply buildconfig/stubs/pygame/transform.pyi]: Done

File: docs/reST/ref/transform.rst [hunk 6]
```diff
@@ -210,6 +210,29 @@ Instead, always begin with the original image and scale to the desired size.)
 
    .. ## pygame.transform.laplacian ##
 
+.. function:: box_blur
+
+   | :sl:`blur a surface using box blur`
+   | :sg:`box_blur(surface, radius, repeat_edge_pixels=True, dest_surface=None) -> Surface`
+
+   Returns the blured surface using box blur algorithm.
```
Comment (oddbookworm) [inline docs/reST/ref/transform.rst:218]: typo, should be "blurred"

File: docs/reST/ref/transform.rst [hunk 7]
```diff
@@ -210,6 +210,29 @@ Instead, always begin with the original image and scale to the desired size.)
...
+   .. versionadded:: 2.1.4
+
+   .. ## pygame.transform.box_blur ##
+
+.. function:: gaussian_blur
+
+   | :sl:`blur a surface using gaussian blur`
+   | :sg:`gaussian_blur(surface, radius, repeat_edge_pixels=True, dest_surface=None) -> Surface`
+
+   Returns the blured surface using gaussian blur algorithm.
```
Comment (oddbookworm) [inline docs/reST/ref/transform.rst:229]: typo, should be "blurred"

**Review by oddbookworm — CHANGES_REQUESTED** [review]
Found a couple typos in the docs. Also, I noticed that the sample image `examples/data/lena.png` appears to be from the internet. I'd rather not deal with copyright infringement accusations, so can you point to the source you got the image from so we can look to see if it's fine to have in the repository?

**Comment by yunline** [comment]
> I'd rather not deal with copyright infringement accusations, so can you point to the source you got the image from so we can look to see if it's fine to have in the repository?

The lena.png is from https://github.com/mikolalysenko/lena.

File: docs/reST/ref/transform.rst (hunk 6 above)
Comment (oddbookworm) [reply docs/reST/ref/transform.rst:218]: this typo still exists

**Review by oddbookworm — CHANGES_REQUESTED** [review]
Ok, I didn't realize this was a standard image in image processing. I'm fine with using it. Although I did stumble upon the [wikipedia](https://en.wikipedia.org/wiki/Lenna) entry, which states that the subject of the image does want the image to be retired and that the use of the image is now controversial. I'll approve it after the typo is fixed, but would like someone else to chime in on whether or not we should be using the image

File: docs/reST/ref/transform.rst (hunk 6 above)
Comment (yunline) [reply docs/reST/ref/transform.rst]: sorry

**Review by Matiiss — APPROVED** [review]
LGTM
I tested with "peppers" (coloured) and "goldhill" images (courtesy of https://links.uwaterloo.ca/Repository.html) and it looked alright so I'm giving a greenlight for this. As for the image, I guess it can be changed to "peppers" since it's apparently the other commonly used test image and it's coloured too, so might as well use that instead I guess (or maybe "monarch").

Test code I used:
```py
import pygame


screen = pygame.display.set_mode((2000, 800))

color = pygame.image.load("my_tests/peppers3.tif").convert()
grayscale = pygame.image.load("my_tests/goldhill1.tif").convert()
print(grayscale.get_size())


# screen.blit(color, (0, 0))
# screen.blit(pygame.transform.box_blur(color, 10), (512, 0))
# screen.blit(pygame.transform.gaussian_blur(color, 10), (1024, 0))

screen.blit(grayscale, (0, 0))
screen.blit(pygame.transform.box_blur(grayscale, 10), (256, 0))
screen.blit(pygame.transform.gaussian_blur(grayscale, 10), (512, 0))

pygame.display.flip()

while not pygame.event.get(pygame.QUIT):
    pass
```

Output:
image
image

**Comment by yunline** [comment]
> I tested with "peppers" (coloured) and "goldhill" images (courtesy of https://links.uwaterloo.ca/Repository.html) and it looked alright so I'm giving a greenlight for this. As for the image, I guess it can be changed to "peppers" since it's apparently the other commonly used test image and it's coloured too, so might as well use that instead I guess (or maybe "monarch").

Sorry for missing your comment. The image is replaced now.

**Review by MyreMylar — APPROVED** [review]
LGTM 👍

**Review by oddbookworm — APPROVED** [review]
LGTM!

**Pull request merged**